projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py

```diff
@@ -0,0 +1,246 @@
+
+from .kwinner_functions import sampled_kwinners, sampled_kwinners2d
+
+
+def update_temperature(m):
+    """Function used to update SampledKWinner modules temperature. This is typically done
+    during training at the beginning of each epoch.
+
+    Call using :meth:`torch.nn.Module.apply` after each epoch if required
+    For example: ``m.apply(update_boost_strength)``
+
+    :param m: KWinner module
+    """
+    if isinstance(m, SampledKWinnersBase):
+        m.update_temperature()
+
+
+class SampledKWinnersBase(nn.Module, metaclass=abc.ABCMeta):
+    """Base SampledKWinners class.
+
+    :param percent_on:
+      k = percent_on * number of input units will be sampled, with replacement, from
+      categorial softmax distribution over the layer activations
+    :type percent_on: float
+
+    :param k_inference_factor:
+      During inference (training=False) we increase percent_on by this factor.
+      percent_on * k_inference_factor must be strictly less than 1.0, ideally much
+      lower than 1.0
+    :type k_inference_factor: float
+
+    :param temperature:
+      temperature to use when passing activations through softmax. Must be >= 1.0
+    :type temperature: float
+
+    :param temperature_decay_rate:
+      reduce the temperature by this much after each epoch, until it hits 1.0.
+    :type temperature_decay_rate: float
+
+    :param eval_temperature:
+      The temperature to use when passing activations through softmax during evaluation.
+      Must be >= 1.0.
+    :type eval_temperature: float
+    """
+
+    def __init__(
+        self,
+        percent_on,
+        k_inference_factor=1.0,
+        temperature=10.0,
+        temperature_decay_rate=0.01,
+        eval_temperature=1.0,
+    ):
+        super(SampledKWinnersBase, self).__init__()
+        assert 0.0 < percent_on < 1.0
+        assert 0.0 < percent_on * k_inference_factor < 1.0
+        assert eval_temperature >= 1.0
+
+        self.percent_on = percent_on
+        self.percent_on_inference = percent_on * k_inference_factor
+        self.temperature = temperature
+        self.temperature_decay_rate = temperature_decay_rate
+        self.eval_temperature = eval_temperature
+        self.learning_iterations = 0
+        self.n = 0
+        self.k = 0
+        self.k_inference = 0
+
+    def _load_from_state_dict(self, *args, **kwargs):
+        super()._load_from_state_dict(*args, **kwargs)
+
+    def extra_repr(self):
+        return (
+            "n={0},"
+            " percent_on={1},"
+            " temperature={2},"
+            " tempreature_decay_rate={3},"
+            " eval_temperature={4}".format(
+                self.n,
+                self.percent_on,
+                self.temperature,
+                self.temperature_decay_rate,
+                self.eval_temperature
+            )
+        )
+
+    def update_temperature(self):
+        new_temp = self.temperature - self.temperature_decay_rate
+        self.temperature = max(1.0, new_temp)
```

Comment:
What's the motivation for a minimum temperature? I get that's if `temperature=1`, you're effectively sampling from the softmax distribution of the unscaled activations. I was thinking there could be a benefit to going lower. As it decays further, you effectively approach a winners take-all.

Reply:
Good point, I'll fix that!

Reply:
Thanks. With that change, there's no subtle bug. When the temperature becomes to low, the softmax distribution has only one non-zero. You can check for this and default to kwinners, like this
```
# Default to kwinners when sampling k is not possible. A low enough temperature
# has only one non-zero in the softmax distribution. However, Multinomial only
# can sample the non-zero probabilites.
if ((probs != 0).sum(dim=1) < k).any():
    # Use kwinners as default; at low temperatures, this is equivalent
    return F.kwinners(x, duty_cycles=None, k=k,
                      boost_strength=0.0,  # no boosting
                      break_ties=False,
                      relu=relu,
                      inplace=inplace)
```
^ And similarly for the 2d version

Reply:
BTW, You should include those comments in the code or modify them just so it's clear why we'd default to regular kwinners.

Reply:
Good point I didn't realize this. I added this to the implementation.